File: DailyChallenges/Scala/LeetCode_129.scala
object Solution {
    def sumNumbers(root: TreeNode): Int = {
        var ans: Int = 0
        def dfs(node: TreeNode, s:Int){
            val _s = s * 10 + node.value
            if (node.left == null && node.right == null) ans += _s
            if (node.left != null) dfs(node.left, _s)
            if (node.right != null) dfs(node.right, _s)
        }
        dfs(root,0)
        ans;
    }
}

File: DailyChallenges/CPP/LeetCode_129.cpp
class Solution {
public:
    int sumNumbers(TreeNode* root) {
        int ans = 0;
        function<void (TreeNode*, int)> dfs = [&](TreeNode* node, int sum){
            int _sum = sum * 10 + node->val;
            if (!node->left && !node->right) { ans += _sum; }
            if (node->left) dfs(node->left, _sum);
            if (node->right) dfs(node->right, _sum);
        };
        dfs(root, 0);
        return ans;
    }
};
